File: hashtable.py
# -*- coding: utf-8 -*-


class HashTable(object):
    def __init__(self):
        self.size = 11
        self.slot = [None] * self.size
        self.data = [None] * self.size

    def put(self, key, data):
        hashvalue = self.hashfunction(key)

        if self.slot[hashvalue] == None:
            self.slot[hashvalue] = key
            self.data[hashvalue] = data
        else:
            if self.slot[hashvalue] == key:
                self.data[hashvalue] = data
            else:
                nextslot = self.rehash(hashvalue)
                while self.slot[nextslot] is not None and self.slot[nextslot] != key:
                    nextslot = self.rehash(nextslot)

                if self.slot[nextslot] == None:
                    self.slot[nextslot] = key
                    self.data[nextslot] = data
                else:
                    self.data[nextslot] = data

    def hashfunction(self, key):
        return key % self.size

    def rehash(self, oldhash):
        return (oldhash+1) % self.size

    def get(self, key):
        startslot = self.hashfunction(key)

        data = None
        stop = False
        found = False
        position = startslot
        while self.slot[position] != None and not stop and not found:
            if self.slot[position] == key:
                found = True
                data = self.data[position]
            else:
                position = self.rehash(position)
                if position == startslot:
                    stop = True

        return data

    def __getitem__(self, key):
        return self.get(key)

    def __setitem__(self, key, data):
        self.put(key, data)

    def __delitem__(self, key):
        self.put(key, None)


if __name__ == '__main__':
    h = HashTable()
    h[54] = 'cat'
    h[26] = 'dog'
    h[93] = 'lion'
    h[17] = 'tiger'
    h[77] = 'bird'
    h[31] = 'cow'
    h[44] = 'goat'
    h[55] = 'pig'
    h[20] = 'chicken'
    print h.slot
    print h.data
    print h[20]
    print h[17]
    h[20] = 'duck'
    print h[20]
    print h.data
    del h[20]
    print h[20]
    print h.data

File: hashtable.cpp
#include <stdio.h>
#include <vector>
#include <string>

using namespace std;

class HashTable
{
public:
    HashTable()
    {
        size = 11;
        slot = vector<int>(size, -1);
        data = vector<string>(size, "");
    }

    ~HashTable()
    {
    }

    void put(int key, string data);
    string get(int key);

public:
    vector<int> slot;
    vector<string> data;

private:
    int hashfunction(int key);
    int rehash(int oldhash);

private:
    int size;
};

int HashTable::hashfunction(int key)
{
    return key % size;
}

int HashTable::rehash(int oldhash)
{
    return (oldhash+1) % size;
}

void HashTable::put(int key, string value)
{
    int hashvalue = hashfunction(key);

    if (slot[hashvalue] == -1)
    {
        slot[hashvalue] = key;
        data[hashvalue] = value;
    }
    else
    {
        if (slot[hashvalue] == key)
        {
            data[hashvalue] = value;
        }
        else
        {
            int nextslot = rehash(hashvalue);
            while (slot[nextslot] != -1 && slot[nextslot] != key)
            {
                nextslot = rehash(nextslot);
            }

            if (slot[nextslot] == -1)
            {
                slot[nextslot] = key;
                data[nextslot] = value;
            }
            else
            {
                data[nextslot] = value;
            }
        }
    }
}

string HashTable::get(int key)
{
    int startslot = hashfunction(key);

    string value = "";
    bool stop = false;
    bool found = false;
    int position = startslot;

    while (slot[position] != -1 && !stop && !found)
    {
        if (slot[position] == key)
        {
            found = true;
            value = data[position];
        }
        else
        {
            position = rehash(position);
            if (position == startslot)
            {
                stop = true;
            }
        }
    }

    return value;
}

int main(void)
{
    HashTable h = HashTable();
    h.put(54, "cat");
    h.put(26, "dot");
    h.put(93, "lion");
    h.put(17, "tiger");
    h.put(77, "bird");
    h.put(31, "cow");
    h.put(44, "goat");
    h.put(55, "pig");
    h.put(20, "chicken");
    for (int i = 0; i < h.slot.size(); i++)
    {
        printf("%d ", h.slot[i]);
    }
    printf("\n");
    for (int i = 0; i < h.data.size(); i++)
    {
        std::string data = h.data[i];
        if (data == "")
            data = "''";
        printf("%s ", data.c_str());
    }
    printf("\n");
    printf("%s\n", h.get(20).c_str());
    printf("%s\n", h.get(17).c_str());
    h.put(20, "duck");
    printf("%s\n", h.get(20).c_str());

    return 0;
}
